micro/service.go

```diff
@@ -0,0 +1,554 @@
+	SchemaVerb
+)
+
+var (
+	// this regular expression is suggested regexp for semver validation: https://semver.org/
+	semVerRegexp      = regexp.MustCompile(`^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$`)
+	serviceNameRegexp = regexp.MustCompile(`^[A-Za-z0-9\-_]+$`)
+)
+
+// Common errors returned by the Service framework.
+var (
+	// ErrConfigValidation is returned when service configuration is invalid
+	ErrConfigValidation = errors.New("validation")
+
+	// ErrVerbNotSupported is returned when invalid [Verb] is used (PING, SCHEMA, INFO, STATS)
+	ErrVerbNotSupported = errors.New("unsupported verb")
+
+	// ErrServiceNameRequired is returned when attempting to generate control subject with ID but empty name
+	ErrServiceNameRequired = errors.New("service name is required to generate ID control subject")
+)
+
+func (s Verb) String() string {
+	switch s {
+	case PingVerb:
+		return "PING"
+	case StatsVerb:
+		return "STATS"
+	case InfoVerb:
+		return "INFO"
+	case SchemaVerb:
+		return "SCHEMA"
+	default:
+		return ""
+	}
+}
+
+// AddService adds a microservice.
+// It will enable internal common services (PING, STATS, INFO and SCHEMA) as well as
+// the actual service handler on the subject provided in config.Endpoint
+// A service name and Endpoint configuration are required to add a service.
+// AddService returns a [Service] interface, allowing service menagement.
+// Each service is assigned a unique ID.
+func AddService(nc *nats.Conn, config Config) (Service, error) {
+	if err := config.valid(); err != nil {
+		return nil, err
+	}
+
+	id := nuid.Next()
+	svc := &service{
+		Config: config,
+		conn:   nc,
+		id:     id,
+	}
+	svc.verbSubs = make(map[string]*nats.Subscription)
+	svc.stats = &Stats{
+		ServiceIdentity: ServiceIdentity{
+			Name:    config.Name,
+			ID:      id,
+			Version: config.Version,
+		},
+	}
+
+	svc.setupAsyncCallbacks()
+
+	// Setup internal subscriptions.
+	var err error
+
+	svc.reqSub, err = nc.QueueSubscribe(config.Endpoint.Subject, QG, func(m *nats.Msg) {
+		svc.reqHandler(&Request{Msg: m})
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	ping := &Ping{
+		Name:    config.Name,
+		ID:      id,
+		Version: config.Version,
+	}
+
+	infoHandler := func(req *Request) {
+		response, _ := json.Marshal(svc.Info())
+		if err := req.Respond(response); err != nil {
+			if err := req.Error("500", fmt.Sprintf("Error handling INFO request: %s", err)); err != nil && config.ErrorHandler != nil {
+				go config.ErrorHandler(svc, &NATSError{req.Subject, err.Error()})
+			}
+		}
+	}
+
+	pingHandler := func(req *Request) {
+		response, _ := json.Marshal(ping)
+		if err := req.Respond(response); err != nil {
+			if err := req.Error("500", fmt.Sprintf("Error handling PING request: %s", err)); err != nil && config.ErrorHandler != nil {
+				go config.ErrorHandler(svc, &NATSError{req.Subject, err.Error()})
```

Comment:
maybe could use `err error` as part of the state from `NATSError`? That way later on could be able to use `errors.Is(err, ErrSlowConsumer)` and handle that case for example

Reply:
I would leave it as it is for now, `ErrSlowConsumer` would be handled internally anyway. We could add it later on if needed.